### PR TITLE
Improve Systolic Array Generator 

### PR DESCRIPTION
**Pull request by calebmkim**
Improved the Systolic Array Generator Frontend so that the PEs can accept a new input every cycle (the PEs use the `pipelined_mult`) primitive. 

It's a bit complicated to explain exactly what the changes are, but the basic idea is that we use the `static repeat` and within the repeat,  we use `static if {idx is between some numbers}` heavily. It's kind of like [this test](https://github.com/cucapra/calyx/blob/master/tests/correctness/static-control/static-mult-dot-product.futil). 

One other note: 
- I added on to the `data_gen` tool and [this python file](https://github.com/cucapra/calyx/blob/17aedc499f6c5bde7bfe7203d08ce66e093d1e47/frontends/systolic-lang/check-output.py#L1) in order to help me test the correctness of the changes I made to the systolic array generator. Should I add test cases that use this to the test suite? The only thing is, I'm not if there's a fud command that you can use for the systolic array generator, or I should just run the python script to get a .futil file, and then run the python script to get the .futil file. 
- Had to change the python library to support static control. 
- There's a probably an increase in resource usage because we need a bunch of `std_lt` and `std_ge` in order to check whether the the index is between certain numbers. But most of these are pretty small so maybe not too concerning. 
- (Edit): the cycle counts that are being showed are beginning to be dominated by writing the output to the PEs to external memory, since that takes `O(ll*tl)` time compared to the actual computation, which takes `O(ll + tl + ld)` time.

### CODE REVIEW

File: frontends/systolic-lang/gen-systolic.py [hunk 1]
```diff
@@ -323,52 +420,58 @@ def counter():
...
-                        py_ast.CompPort(py_ast.CompVar(f"left_{r}_{c}"), "out"),
-                    ),
-                ],
-                out_connects=[],
+    control_stmts = []
+    incr_stmts = [py_ast.Enable("incr_idx")]
+    for r in range(left_length):
+        for c in range(top_length):
+            # build 4 if stmts for the 4 schedules that we need to account for
+            input_mem_updates = execute_if_between(
```
Comment (rachitnigam) [inline frontends/systolic-lang/gen-systolic.py:428]: This is cool! It's like we're building a DSL of "instructions" (like `execute_if_between`) which then gets turned into hardware. You can imagine that in the future, we could build a systolic array DSL with literally these operations and have users program in it!

**Review by rachitnigam — APPROVED** [review]
This is looking great! I particularly appreciate how the changes are moving us from a "script" a more of a domain-specific language (DSL). Responses to your questions

> Should I add test cases that use this to the test suite? The only thing is, I'm not if there's a fud command that you can use for the systolic array generator, or I should just run the python script to get a .futil file, and then run the python script to get the .futil file.

That sounds great! `fud` can transform a systolic configuration file (which is just a JSON file with the dimensions) into a `.futil` file using `fud e <file> --from systolic ...`.

> There's a probably an increase in resource usage because we need a bunch of std_lt and std_ge in order to check whether the the index is between certain numbers. But most of these are pretty small so maybe not too concerning.

Yup, matches my intuition as well! Plus, it means that we can support matrices with arbitrary dimension sizes (as long as their shared dimension matches)

> (Edit): the cycle counts that are being showed are beginning to be dominated by writing the output to the PEs to external memory, since that takes O(ll*tl) time compared to the actual computation, which takes O(ll + tl + ld) time.

Hm, can explain that a bit more? The PEs all store the outputs and should be able to transfer the data in parallel? Is it because we're using a sequential memory to write the values back?

We'll need to address this if we want a real competitive baseline. Thoughts and possible solutions:
1. Pipeline the output writes. Instead of writing the outputs when all the PEs are done, we can start writing the values from the top PEs first. This will only give us a constant factor speedup so maybe not worth it. In real systolic arrays, the reading of the outputs is pipelined with the next execution to "hide the latency" of the memory operation. Alternatively, people build systolic arrays with a "weight-stationary" dataflow (instead of the output stationary dataflow that we implement) to make the output reads cheaper. @sampsyo might have more thoughts on this
2. We ignore the the latency of reading the outputs. This is only kind of okay and something we shouldn't do if we build a weight stationary systolic array too.
3. Going back to the point of latency hiding: in real workloads, you rarely do one matrix multiply at a time. Instead you're doing tons of operations back-to-back and want to pipeline their execution. Our benchmark (running one matrix-multiply) gives you the worst cast bound on the performance.

Another big thing to take from 1-3 is that accelerator performance is hard to evaluate fairly. Cycle count is just one metric and there are many others that give a more realistic picture when it comes to system-level concerns. As I mentioned in our slack, read Sophia Shao's dissertation to get a better sense of the system-level concerns. It might not yield anything actionable for the paper but is definitely useful to know.

**Comment by calebmkim** [comment]
So to expand on the problem with the cycle count, the fundamental problem is that the output is currently just a single external memory, which means it can only support one write in a given cycle. 

I think pipelining the outputs of the writes to memory could be feasible if we change the representation of memory to be a bunch of 1 dimensional memories (like how the inputs are currently stored).

**Comment by rachitnigam** [comment]
Hm, we should at least bank the memory to allow for parallel writes in the same way the input memories are banked.

**Comment by calebmkim** [comment]
I added some correctness cases. 

I think I'll merge this, and work on the banking the output memories in a separate PR, given that this PR is already pretty big. I'm optimistic that doing this will dramatically reduce the cycle count, since on each "row" of the output array there is (at most) one entry in the row whose value is going to be "finished" at the end of each cycle.

**Comment by sampsyo** [comment]
Awesome! To distill the discussion about the cost of writing out results, it seems like there are two issues:

1. Overlapping the SA computation with the final output-writing.
2. The output is a single-ported memory, so it takes at least n cycles to write n values.

I have one super high-level question about item 1: is this also a problem while we're reading the inputs? I think the answer is "no" because the way we're running the SA only requires reading a single value from the (single-ported) input memories per cycle, in the steady state. Is that right?

It's definitely a tricky issue, and there is an argument for factoring out the I/O and measuring it separately from the core computation. On the other hand, it does seem most fair to imagine a banked I/O interface, which would be slightly challenging but also rewarding in the sense that it would hopefully highlight the real-world strengths or weaknesses of a particular systolic array setup without idealizing anything… on that latter path, it might be useful to google around for a couple of papers (or open-source projects?) about output-stationary systolic GEMMs to see what they do about output? There's no shame in stealing ideas for this purpose!

**Comment by calebmkim** [comment]
> I have one super high-level question about item 1: is this also a problem while we're reading the inputs? I think the answer is "no" because the way we're running the SA only requires reading a single value from the (single-ported) input memories per cycle, in the steady state. Is that right?

Yep, that's right. And just to clarify even further, only the left and upper "edges" of the SA actually need to read memory: the rest of the PEs get the data because it's passed along from one PE to the next. 

I think I actually have decent idea of how to implement the output with banked memories (and writes during the SA computation), although there may be something I'm not thinking of... in any case, I can always look online for inspiration.

**Comment by rachitnigam** [comment]
One possible option here is recreating the "systolic flow" of the inputs for the outputs as well: Each PE contains pushes the value of the output to the PE on the left when its done and again, because of the schedule, we know that there will be no conflicts. This is only useful inasmuch as it reduces the potential length of the output wires since they just need to send values back to their neighbors

**Pull request merged**